ruby/lib/google/protobuf/ffi/descriptor.rb

```diff
@@ -0,0 +1,765 @@
+                clear_internal(field_descriptor)
+              end
+              if field_descriptor.type == :enum
+                define_method("#{field_name}_const") do
+                  if field_descriptor.repeated?
+                    return_value = []
+                    get_field(field_descriptor).send(:each_msg_val) do |msg_val|
+                      return_value << msg_val[:int32_val]
+                    end
+                    return_value
+                  else
+                    message_value = Google::Protobuf::FFI.get_message_value @msg, field_descriptor
+                    message_value[:int32_val]
+                  end
+                end
+              end
+              if !field_descriptor.repeated? and field_descriptor.wrapper?
+                define_method("#{field_name}_as_value") do
+                  get_field(field_descriptor, unwrap: true)
+                end
+                define_method("#{field_name}_as_value=") do |value|
+                  if value.nil?
+                    clear_internal(field_descriptor)
+                  else
+                    index_assign_internal(value, field_descriptor: field_descriptor, wrap: true)
+                  end
+                end
+              end
+              if field_descriptor.has_presence?
+                if field_descriptor.sub_message? or field_descriptor.send(:real_containing_oneof).nil? or
+                  Google::Protobuf::FFI.message_def_syntax(Google::Protobuf::FFI.get_containing_message_def(field_descriptor)) == :Proto2
+                  define_method("has_#{field_name}?") do
+                    Google::Protobuf::FFI.get_message_has(@msg, field_descriptor)
+                  end
+                end
+              end
+            end
+          end
+
+          @oneof_field_names = []
+
+          @descriptor.each_oneof do |oneof_descriptor|
+            field_name = oneof_descriptor.name.to_sym
+            @oneof_field_names << field_name
+            unless instance_methods(true).include?(field_name)
+              define_method(field_name) do
+                field_descriptor = Google::Protobuf::FFI.get_message_which_oneof(@msg, oneof_descriptor)
+                if field_descriptor.nil?
+                  return
+                else
+                  return field_descriptor.name.to_sym
+                end
+              end
+              define_method("clear_#{field_name}") do
+                field_descriptor = Google::Protobuf::FFI.get_message_which_oneof(@msg, oneof_descriptor)
+                unless field_descriptor.nil?
+                  clear_internal(field_descriptor)
+                end
+              end
+              define_method("has_#{field_name}?") do
+                !Google::Protobuf::FFI.get_message_which_oneof(@msg, oneof_descriptor).nil?
+              end
+            end
+          end
+
+          private
+          # Implementation details below are subject to breaking changes without
```

Comment:
If these are truly private, do we need this warning?

Reply:
I don't feel strongly about this, but since we'd gone to extraordinary measures in the CRuby implementation to frustrate direct calls from Ruby for various methods, I opted to be more explicit. If you want me to remove them from across the PR, let me know.